### DISCUSSION TITLE
using re timers properly when multithreaded

### DISCUSSION THREAD

**Discussion by fAuernigg** [baresip/re#139] [Q&A] [answered]
"tmr_start" adds a tmr element to the global tmr list.
https://github.com/baresip/re/blob/ae645198c3b48991ef5947aad040195d80810528/src/tmr/tmr.c#L302

Is it safe to call tmr_start from a thread different to the libre main thread?

What happens, if the thread adding an element to the timer list is interrupted by the libre main thread. Could it crash or have other impacts?
Is there a way to lock the tmr list during modification?

**Comment by sreimers** [comment, marked as answer]
For complex multithread scenarios you can use [mqueue](https://github.com/baresip/re/blob/master/src/mqueue/mqueue.c).

**Reply by cspiel1** [reply]
Thanks for this hint! We realized now that:
- `tmr_start` (and most likely also `tmr_cancel`) don't have to be thread-safe for baresip. They are only called from `re_main` thread.
- `mqueue` supports very handy passing control to the re_main thread. We will use this in our application to pass the `tmr_start` invocation to the re_main thread.
